x-pack/platform/plugins/shared/onechat/server/routes/esql_tools.ts

```diff
@@ -0,0 +1,227 @@
+  getInternalServices,
+  logger,
+}: RouteDependencies) {
+  const wrapHandler = getHandlerWrapper({ logger });
+
+  router.get(
+    {
+      path: '/api/chat/tools/esql/{id}',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
+      },
+      validate: {
+        params: schema.object({
+          id: schema.string(),
+        }),
+      },
+    },
+    wrapHandler(async (ctx, request, response) => {
+      try {
+        const { esql: esqlToolClientService } = getInternalServices();
+        const client = await esqlToolClientService.getScopedClient({ request });
+        const tool = await client.get(request.params.id);
+
+        return response.ok({
+          body: {
+            tool,
+          },
+        });
+      } catch (error) {
+        throw error;
+      }
+    })
+  );
+
+  router.get(
+    {
+      path: '/api/chat/tools/esql',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
+      },
+      validate: false,
+    },
+    wrapHandler(async (ctx, request, response) => {
+      try {
+        const { esql: esqlToolClientService } = getInternalServices();
+        const client = await esqlToolClientService.getScopedClient({ request });
+
+        const tools = await client.list();
+
+        return response.ok({
+          body: {
+            tools,
+          },
+        });
+      } catch (error) {
+        throw error;
+      }
+    })
+  );
+
+  router.post(
+    {
+      path: '/api/chat/tools/esql',
+      security: {
+        authz: { requiredPrivileges: [apiPrivileges.readOnechat] },
+      },
+      validate: {
+        body: schema.object({
+          id: schema.string(),
+          description: schema.string(),
+          query: schema.string(),
+          params: schema.recordOf(
+            schema.string(),
+            schema.object({
+              type: schema.string(),
+              description: schema.string(),
+            })
+          ),
+          meta: schema.object({
+            tags: schema.arrayOf(schema.string(), { defaultValue: [] }),
+          }),
+        }),
+      },
+    },
+    wrapHandler(async (ctx, request, response) => {
+      try {
+        const { esql: esqlToolService } = getInternalServices();
+        const client = await esqlToolService.getScopedClient({ request });
+
+        const now = new Date();
+
+        const tool: EsqlToolCreateRequest = {
+          id: request.body.id || 'esql',
+          description: request.body.description || 'Tool for executing ESQL queries',
+          query: request.body.query,
+          params: request.body.params || {},
+          meta: {
+            tags: request.body.meta.tags || [],
+            providerId: esqlToolProviderId,
+          },
+          created_at: now.toISOString(),
+          updated_at: now.toISOString(),
+        };
+
+        await client.create(tool);
+
+        return response.ok({
+          body: {
+            tool: { ...request.body },
```

Comment:
IMO, the client.create() signature should be updated to read and return the saved tool object from storage.

Returning the original request body feels like bad practice, it can hide issues and make debugging harder.